Microsoft.Identity.Web/Client/TokenCacheProviders/Redis/MSALAppRedisTokenCacheProvider.cs

```diff
@@ -0,0 +1,137 @@
+﻿/************************************************************************************************
+The MIT License (MIT)
+
+Copyright (c) 2015 Microsoft Corporation
+
+Permission is hereby granted, free of charge, to any person obtaining a copy
+of this software and associated documentation files (the "Software"), to deal
+in the Software without restriction, including without limitation the rights
+to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+copies of the Software, and to permit persons to whom the Software is
+furnished to do so, subject to the following conditions:
+
+The above copyright notice and this permission notice shall be included in all
+copies or substantial portions of the Software.
+
+THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+SOFTWARE.
+***********************************************************************************************/
+
+using Microsoft.AspNetCore.Authentication.AzureAD.UI;
+using Microsoft.AspNetCore.Http;
+using Microsoft.Extensions.Caching.Distributed;
+using Microsoft.Extensions.Options;
+using Microsoft.Identity.Client;
+using System;
+
+namespace Microsoft.Identity.Web.Client.TokenCacheProviders
+{
+    /// <summary>
+    /// An implementation of token cache for both Confidential and Public clients backed by Redis Cache.
+    /// When cached data is distributed, the data:
+    /// Is coherent(consistent) across requests to multiple servers.
+    /// Survives server restarts and app deployments.
+    /// Doesn't use local memory.
+    /// </summary>
+    /// <seealso cref="https://aka.ms/msal-net-token-cache-serialization"/>
+    public class MSALAppRedisTokenCacheProvider : IMSALAppTokenCacheProvider
+    {
+        /// <summary>
+        /// The application cache key
+        /// </summary>
+        internal string AppCacheId;
+
+        /// <summary>
+        /// The backing IDistributedCache instance
+        /// </summary>
+        internal IDistributedCache distributedCache;
+
+        private readonly MSALRedisTokenCacheOptions CacheOptions;
+
+        /// <summary>
+        /// The App's whose cache we are maintaining.
+        /// </summary>
+        private readonly string AppId;
+
+        public MSALAppRedisTokenCacheProvider(IDistributedCache cache,
+            MSALRedisTokenCacheOptions option,
+            IOptionsMonitor<AzureADOptions> azureAdOptionsAccessor)
+        {
```

Comment:
missing validation for some of the input parameters:
`if (cache == null) throw new ArgumentNullException(nameof(cache));`
and 
`if (azureAdOptionsAccessor == null) throw new ArgumentNullException(nameof(azureAdOptionsAccessor));`